### PR TITLE
Add built-in support for spellchecking

### PR DESCRIPTION
**Pull request by nautatva**
Built in support for spell check suggestions. This fixes https://github.com/sindresorhus/electron-context-menu/issues/11.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#11: Built in support for spell check suggestions](https://issuehunt.io/repos/60405378/issues/11)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

**Comment by nautatva** [comment]
This will also solve https://github.com/sindresorhus/caprine/issues/14 which was due and waiting for this PR.

File: fixture.js [hunk 1]
```diff
@@ -34,7 +34,7 @@ contextMenu({
 			type: 'separator'
 		}
 	],
-	append: () => {},
+	append: () => { },
```
Comment (sindresorhus) [inline fixture.js]: Don't do unrelated changes.

File: fixture-menu.js [hunk 2]
```diff
@@ -33,5 +33,10 @@ contextMenu({
 (async () => {
 	await app.whenReady();
 
-	await (new BrowserWindow()).loadFile(path.join(__dirname, 'fixture.html'));
+	await (new BrowserWindow({
+		webPreferences: {
+			nodeIntegration: true,
+			spellcheck: true
```
Comment (sindresorhus) [inline fixture-menu.js]: The fact that this is needed needs to be documented.

File: index.d.ts [hunk 3]
```diff
@@ -217,7 +217,8 @@ declare namespace contextMenu {
 		- `showSaveImageAs`
 		- `showInspectElement`
 		- `showServices`
-
+		- `dictionary`
+
```
Comment (sindresorhus) [inline index.d.ts]: Don't include whitespace.

File: index.js [hunk 4]
```diff
@@ -173,7 +173,24 @@ const create = (win, options) => {
 
 		const shouldShowInspectElement = typeof options.showInspectElement === 'boolean' ? options.showInspectElement : isDev;
 
+		const dictSuggestions = [];
+		for (let index = 0; index < props.dictionarySuggestions.length; index++) {
```
Comment (sindresorhus) [inline index.js]: `Array#map` would be more appropriate here.

File: index.js [hunk 5]
```diff
@@ -173,7 +173,24 @@ const create = (win, options) => {
 
 		const shouldShowInspectElement = typeof options.showInspectElement === 'boolean' ? options.showInspectElement : isDev;
 
+		const dictSuggestions = [];
```
Comment (sindresorhus) [inline index.js]:
```suggestion
		const dictionarySuggestions = [];
```

File: index.js [hunk 6]
```diff
@@ -173,7 +173,24 @@ const create = (win, options) => {
...
+		for (let index = 0; index < props.dictionarySuggestions.length; index++) {
+			dictSuggestions.push(
+				{
+					id: 'dictSuggestions',
+					label: props.dictionarySuggestions[index],
+					visible: props.isEditable && hasText && props.misspelledWord,
```
Comment (sindresorhus) [inline index.js]: The `props.isEditable && hasText` check should be before the loop. We should not even do any work if it cannot be used.

File: package.json [hunk 7]
```diff
@@ -47,4 +47,4 @@
 			"browser"
 		]
 	}
-}
+}
```
Comment (sindresorhus) [inline package.json]: ^

File: readme.md [hunk 8]
```diff
@@ -185,7 +185,8 @@ The following options are ignored when `menu` is used:
 - `showSaveImageAs`
 - `showInspectElement`
 - `showServices`
-
+- `dictionary`
+
```
Comment (sindresorhus) [inline readme.md]: Trailing whitespace

File: readme.md [hunk 9]
```diff
@@ -200,6 +201,7 @@ Default actions:
 - `copyLink`
 - `inspect`
 - `services`
+- `dictionary`
```
Comment (sindresorhus) [inline readme.md]: I don't think `dictionary` is the correct word to use here.

**Comment by sindresorhus** [comment]
Make sure it looks exactly like: https://github.com/sindresorhus/electron-context-menu/issues/11#issuecomment-582300971 I think you've missed some things.

**Comment by sindresorhus** [comment]
This issue has a high bounty, so it's expected that you put a *lot* of effort into this PR ;)

File: fixture-menu.js (hunk 2 above)
Comment (shahan312) [reply fixture-menu.js]: Seems like this is needed for the `searchWithGoogle` option being added. Would it be possible to implement a callback for the menu option, and avoid the `require` call? I think its very important to have a workaround for the nodeIntegration requirement, as there is no way we would ever turn that on in our usage, for security reasons.

**Comment by shahan312** [comment]
Seems like the `nodeIntegration: true` is needed for the `searchWithGoogle` option being added. Would it be possible to implement a callback alternative for the menu option, and to avoid the require call? I think it is very important to have a workaround for the `nodeIntegration` requirement, as there is no way we would ever turn that on in our usage, for security reasons.

**Comment by nautatva** [comment]
- [x] Spelling suggestions
- [ ] Ignore spelling
- [x] Correct spelling Automatically
- [x] Learn spelling
- [x] search with google

Implemented these things in this commit as from the figure in https://github.com/sindresorhus/electron-context-menu/issues/11#issuecomment-582300971. 

The bottom-most options seem more of a system-wide setting change rather than a app-context. Not sure if we can implement this on a non-system app. Even MS-word doesn't have these options.
I was stuck at implementing the ignore spelling part. Can you help me with that please @sindresorhus

**Comment by nautatva** [comment]
I re-checked the `searchWithGoogle` option and it works even without `nodeIntegration: true`. Can you recheck if something else isn't being missed out.

File: index.js (hunk 4 above)
Comment (nautatva) [reply index.js]: Performance Analysis For loops performs faster than map or foreach as number of elements in a array increases. https://stackoverflow.com/a/56275105

**Comment by shahan312** [comment]
> I re-checked the `searchWithGoogle` option and it works even without `nodeIntegration: true`. Can you recheck if something else isn't being missed out.

This makes sense, since the context menu is setup in the main process.

File: index.js (hunk 4 above)
Comment (sindresorhus) [reply index.js]: That is irrelevant for such a small array. You would have to run through tens of thousands of elements for that to matter.

**Comment by sindresorhus** [comment]
> The bottom-most options seem more of a system-wide setting change rather than a app-context. 

Yup. Those are not relevant here. And not possible to implement.

**Comment by sindresorhus** [comment]
> I was stuck at implementing the ignore spelling part. Can you help me with that please @sindresorhus

What are you stuck with specifically? What have you tried?

File: index.d.ts [hunk 10]
```diff
@@ -14,9 +14,14 @@ declare namespace contextMenu {
 		/**
 		The placeholder `{selection}` will be replaced by the currently selected text.
 
-		@default 'Look Up “{selection}”'
+		@default 'Look Up "{selection}"'
```
Comment (sindresorhus) [inline index.d.ts]: Don't do unrelated changes.

File: index.js [hunk 11]
```diff
@@ -51,6 +51,14 @@ const create = (win, options) => {
 					}
 				}
 			}),
+			searchWithGoogle: decorateMenuItem({
+				id: 'searchWithGoogle',
+				label: 'Search with Google',
+				visible: hasText,
+				click() {
+					require('electron').shell.openExternal('https://www.google.com/search?q=' + props.selectionText);
```
Comment (sindresorhus) [inline index.js]: Electron is already required. Also use template literal interpolation.

**Comment by sindresorhus** [comment]
Don't forget to document all the menu items and options, in both the readme and the TypeScript file. I see some missing things and inconsistencies. The readme and TypeScript file should be in sync.

**Comment by nautatva** [comment]
Had opened a electron issue for the same: https://github.com/electron/electron/issues/22362. I think ignore should be out of question here.

**Comment by nautatva** [comment]
Any updates for changes?

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/electron-context-menu/pull/94/files#r378506080 is not fixed.

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/electron-context-menu/pull/94#issuecomment-589937031 is not fixed.

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/electron-context-menu/pull/94#discussion_r382899659 is not fixed.

**Comment by sindresorhus** [comment]
Please spend some more time improving the docs and perfecting this PR :)

**Comment by nautatva** [comment]
Have a look at the changes in this commit
## Docs
**spellcheck needed documented**
- [x] readme
- [x] typescript

**showSearchWithGoogle**
- [x] readme
- [x] typescript

**searchWithGoogle**
- [x] readme
- [x] typescript

**spellCheck**
- [x] readme
- [x] typescript

**correctAutomatically**
- [x] readme
- [x] typescript

**learnSpelling**
- [x] readme
- [x] typescript

## Other changes
- [x] used array#map instead of for loop
- [x] encoding the url

File: index.js [hunk 12]
```diff
@@ -56,8 +56,8 @@ const create = (win, options) => {
 				label: 'Search with Google',
 				visible: hasText,
 				click() {
-					const query = 'https://www.google.com/search?q=';
-					electron.shell.openExternal(query + props.selectionText);
+					const url = new URL('https://www.google.com/search?q=' + props.selectionText).href;
```
Comment (sindresorhus) [inline index.js]: This is still not correct. I asked for the query string to be properly encoded. You do that with `URL#searchParams`.

**Comment by nautatva** [comment]
Is this okay? Haven't work with this before.

File: index.js [hunk 13]
```diff
@@ -56,7 +56,9 @@ const create = (win, options) => {
 				label: 'Search with Google',
 				visible: hasText,
 				click() {
-					const url = new URL('https://www.google.com/search?q=' + props.selectionText).href;
+					const queryParams = {q: props.selectionText};
+					const queryString = new URLSearchParams(queryParams).toString();
+					const url = new URL('https://www.google.com/search?' + queryString).href;
```
Comment (sindresorhus) [inline index.js]: This is not what I asked for. I asked you to use `URL#searchParams`.  `const url = new URL(…); url.searchParams.set(…);`.

Comment (sindresorhus) [reply index.js]: And use `url.toString()`.

Comment (nautatva) [reply index.js]: Did this in the following commit, but am curious, why can't we use `const url = new URL('https://www.google.com/search?' + new URLSearchParams(queryParams).toString()).href;`

Is it just because of readability or it due to some other reason as well?

**Comment by nautatva** [comment]
Does this work?

File: index.js [hunk 14]
```diff
@@ -152,6 +162,24 @@ const create = (win, options) => {
 					});
 				}
 			}),
+			correctAutomatically: decorateMenuItem({
```
Comment (sindresorhus) [inline index.js:165]: Where did you get this menu item from? I don't see it when misspelling a word in Safari.

File: index.js [hunk 15]
```diff
@@ -152,6 +162,24 @@ const create = (win, options) => {
...
+				id: 'correctAutomatically',
+				label: 'Correct Spelling Automatically',
+				visible: props.isEditable && hasText && props.misspelledWord && props.dictionarySuggestions.length,
```
Comment (sindresorhus) [inline index.js:168]:
```suggestion
				visible: props.isEditable && hasText && props.misspelledWord && props.dictionarySuggestions.length > 0,
```

File: index.js [hunk 16]
```diff
@@ -173,10 +201,44 @@ const create = (win, options) => {
...
+		}
+
+		let dictionarySuggestions = [];
+		if (hasText && props.misspelledWord && props.dictionarySuggestions.length > 0) {
+			dictionarySuggestions = props.dictionarySuggestions.map(word);
+		} else {
+			dictionarySuggestions.push(
+				{
+					id: 'dictionarySuggestions',
+					label: 'No guesses available',
```
Comment (sindresorhus) [inline index.js:223]:
```suggestion
					label: 'No Guesses Found',
```

File: index.js (hunk 14 above)
Comment (nautatva) [reply index.js:165]: It was in the image, as on comment:
https://github.com/sindresorhus/electron-context-menu/issues/11#issuecomment-582300971

File: index.js (hunk 13 above)
Comment (sindresorhus) [reply index.js]: You technically could, but what if you forget to put the `?` at the end. I prefer using dedicated APIs for formatting whenever possible to prevent mistakes.

**Comment by sindresorhus** [comment]
Yup, looks good now 👍

**Comment by sindresorhus** [comment]
@nautatva Can you do a PR to Caprine now to get the bounty there too?

**Comment by nautatva** [comment]
Done!
